In [19]:
%load_ext autoreload
%autoreload 2
import sys
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
from tqdm import tqdm
import wandb

# import constants
from constants import *

# Add the parent directory to the system path
sys.path.append("..")

from utils import load_data_multilabel_pre_split , create_dataloaders, initialization_using_word2vec, create_subset_dataloader, get_micro_metrics, create_vocabulary_label_pre_split, create_vocabulary
from HAN_model import HierarchicalAttentionNetwork

from metrics import *
from train import train

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [11]:
train_dataloader, valid_dataloader, vocab_size = create_dataloaders()

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True
load_data.started...
load_data_multilabel_new.data_path: ../datasets/data/train_50_eamc.csv
load_data.ended...
load_data.started...
load_data_multilabel_new.data_path: ../datasets/data/dev_50_eamc.csv
load_data.ended...
shuffled training data


In [4]:
model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)

In [5]:
model = initialization_using_word2vec(model)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


In [6]:
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.000005)

### first training script, without initialization.
this training script has one major fault: we forgot to reset gradients at each minibatch! Keeping here for reference

In [18]:
# train one epoch
from tqdm import tqdm

losses = []
for i, (x, y) in tqdm(enumerate(dataloader)):
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())
print('done 1 epoch', 'loss:', np.array(losses).mean())

1it [00:04,  4.10s/it]

iter 0 loss 52.44690704345703


11it [00:32,  2.65s/it]

iter 10 loss 36.18838847767223


21it [00:58,  2.61s/it]

iter 20 loss 33.24757476080032


31it [01:24,  2.60s/it]

iter 30 loss 32.721970650457564


41it [01:50,  2.62s/it]

iter 40 loss 32.93682638028773


51it [02:16,  2.61s/it]

iter 50 loss 31.958648382448683


61it [02:42,  2.64s/it]

iter 60 loss 31.103213200803662


71it [03:09,  2.64s/it]

iter 70 loss 30.552288727021555


81it [03:40,  2.73s/it]

iter 80 loss 30.282272644984868


91it [04:06,  2.68s/it]

iter 90 loss 29.822059254069906


101it [04:39,  3.12s/it]

iter 100 loss 29.609745006750124


111it [05:09,  2.71s/it]

iter 110 loss 29.421181876380164


121it [05:35,  2.64s/it]

iter 120 loss 29.341002503702462


131it [06:02,  2.65s/it]

iter 130 loss 29.239765793312596


141it [06:29,  2.82s/it]

iter 140 loss 29.16025330834355


151it [07:03,  3.36s/it]

iter 150 loss 29.132182456010224


161it [07:37,  3.40s/it]

iter 160 loss 28.99819346836635


171it [08:10,  3.21s/it]

iter 170 loss 28.91697088319656


181it [08:39,  2.72s/it]

iter 180 loss 28.848535927619725


191it [09:05,  2.62s/it]

iter 190 loss 28.810952900591946


201it [09:35,  2.91s/it]

iter 200 loss 28.856003946332788


211it [10:08,  3.71s/it]

iter 210 loss 28.91458556550374


221it [10:43,  3.70s/it]

iter 220 loss 29.132748237022987


231it [11:18,  3.33s/it]

iter 230 loss 29.444872472193335


241it [11:53,  3.25s/it]

iter 240 loss 29.84288608978398


251it [12:22,  2.78s/it]

iter 250 loss 30.278502186930986


252it [12:25,  2.96s/it]


RuntimeError: shape '[32, 2500, 100]' is invalid for input of size 500000

In [22]:
# train one epoch
from tqdm import tqdm

losses = []
for i, (x, y) in tqdm(enumerate(dataloader)):
    optimizer.zero_grad()
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())
print('done 1 epoch', 'loss:', np.array(losses).mean())

1it [00:04,  4.81s/it]

iter 0 loss 75.09092712402344


11it [00:37,  3.11s/it]

iter 10 loss 49.521841916170985


21it [01:04,  2.70s/it]

iter 20 loss 43.32634353637695


31it [01:32,  2.67s/it]

iter 30 loss 40.6211912093624


41it [02:02,  2.84s/it]

iter 40 loss 39.0463952320378


51it [02:28,  2.65s/it]

iter 50 loss 37.02092331531001


61it [02:57,  3.18s/it]

iter 60 loss 35.387808909181686


71it [03:29,  2.70s/it]

iter 70 loss 34.229467929249076


81it [03:56,  2.93s/it]

iter 80 loss 33.58679429984387


91it [04:24,  2.55s/it]

iter 90 loss 32.963436902224366


101it [04:49,  2.50s/it]

iter 100 loss 32.51199756282391


111it [05:14,  2.49s/it]

iter 110 loss 32.14766309927176


121it [05:39,  2.52s/it]

iter 120 loss 31.86359189561576


131it [06:05,  2.58s/it]

iter 130 loss 31.611873408310284


141it [06:31,  2.57s/it]

iter 140 loss 31.395271436542483


151it [06:57,  2.60s/it]

iter 150 loss 31.239010918219357


161it [07:23,  2.58s/it]

iter 160 loss 31.03779889930109


171it [07:49,  2.58s/it]

iter 170 loss 30.91952669690227


181it [08:15,  2.60s/it]

iter 180 loss 30.828073501586914


191it [08:41,  2.60s/it]

iter 190 loss 30.74624914398992


201it [09:07,  2.66s/it]

iter 200 loss 30.774566384690317


211it [09:33,  2.63s/it]

iter 210 loss 30.809242781869607


221it [10:00,  2.71s/it]

iter 220 loss 30.96526320271902


231it [10:27,  2.65s/it]

iter 230 loss 31.193378225549473


241it [10:54,  2.65s/it]

iter 240 loss 31.55847540732736


251it [11:20,  2.64s/it]

iter 250 loss 31.937302692002984


253it [11:23,  2.70s/it]

done 1 epoch loss: 32.009809041682914


Let's evaluate the model after 1 epoch

In [61]:
preds = []
actuals = []
with torch.no_grad():
    for x, y in tqdm(valid_dataloader):
        y_hat = model(x)
        probs = F.softmax(y_hat, dim=1)

        label = torch.argmax(probs, dim=1)
        label_y = torch.argmax(y, dim=1)

        preds.append(list(label.numpy()))
        actuals.append(list(label_y.numpy()))

        

100%|██████████| 50/50 [00:46<00:00,  1.06it/s]


In [63]:
preds = np.concatenate(preds)

In [64]:
actuals = np.concatenate(actuals)

In [65]:
acc, prec, rec, f1 = all_macro(preds, actuals)
acc, prec, rec, f1

(0.3801787164906271,
 0.057303783519039336,
 0.07258064516128919,
 0.06404379062606823)

In [66]:
precision, recall = calculate_precision_recall(preds, actuals)
print(precision, recall)

0.4259376986649714 0.7745664739884393


In [70]:
from sklearn.metrics import precision_score, recall_score

# Assume y_true and y_pred are your actual and predicted labels respectively
# y_true = ...
# y_pred = ...

precision = precision_score(preds, actuals, average='macro')  # for multi-class classification
recall = recall_score(preds, actuals, average='macro')  # for multi-class classification

print(f"Precision: {precision}")
print(f"Recall: {recall}")

Precision: 0.024237049531897863
Recall: 0.01719362244931067


/Users/hungryfoolish/anaconda3/envs/bd4h/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/hungryfoolish/anaconda3/envs/bd4h/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [71]:
precision = precision_score(preds, actuals, average='micro')  # for multi-class classification
recall = recall_score(preds, actuals, average='micro')  # for multi-class classification

print(f"Precision: {precision}")
print(f"Recall: {recall}")

Precision: 0.22822631913541006
Recall: 0.22822631913541006


### Training one epoch with smart initialization

In [8]:
# train one epoch
from tqdm import tqdm

losses = []
for i, (x, y) in tqdm(enumerate(train_dataloader)):
    optimizer.zero_grad()
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())
print('done 1 epoch', 'loss:', np.array(losses).mean())

1it [00:05,  5.59s/it]

iter 0 loss 22.66522979736328


11it [00:37,  3.26s/it]

iter 10 loss 17.240951017899945


21it [01:05,  2.71s/it]

iter 20 loss 16.972276778448197


31it [01:32,  2.68s/it]

iter 30 loss 17.674793612572454


41it [01:59,  2.71s/it]

iter 40 loss 18.54887045883551


51it [02:27,  2.74s/it]

iter 50 loss 18.660027148676853


61it [02:56,  3.13s/it]

iter 60 loss 18.769797246964252


71it [03:22,  2.66s/it]

iter 70 loss 18.92314792686785


81it [03:50,  2.69s/it]

iter 80 loss 19.220837534209828


91it [04:20,  2.96s/it]

iter 90 loss 19.411412794511396


101it [04:48,  2.77s/it]

iter 100 loss 19.68570909405699


111it [05:21,  3.78s/it]

iter 110 loss 19.983266967910904


121it [05:54,  3.05s/it]

iter 120 loss 20.294265494858923


131it [06:23,  2.88s/it]

iter 130 loss 20.604145006369087


141it [06:55,  3.09s/it]

iter 140 loss 20.911784801077335


151it [07:23,  2.80s/it]

iter 150 loss 21.23349266810133


161it [07:50,  2.75s/it]

iter 160 loss 21.453838336541786


171it [08:18,  2.78s/it]

iter 170 loss 21.71918661552563


181it [08:45,  2.75s/it]

iter 180 loss 21.973330429245753


191it [09:13,  2.81s/it]

iter 190 loss 22.211702636399195


201it [09:41,  2.75s/it]

iter 200 loss 22.508781490041248


211it [10:09,  2.78s/it]

iter 210 loss 22.759959107891643


221it [10:37,  2.80s/it]

iter 220 loss 23.062788950372067


231it [11:05,  2.81s/it]

iter 230 loss 23.33391928569579


241it [11:33,  2.81s/it]

iter 240 loss 23.66225935908274


251it [12:01,  2.90s/it]

iter 250 loss 24.02995038887419


253it [12:04,  2.86s/it]

done 1 epoch loss: 24.120068244783305


loss starts out much lower (less than 40% of the model without initialization)
However loss didn't reduce after 1 epoch. To look into:
- gradient clipping
- learning rate scheduler

In [11]:
logger = SummaryWriter("./logs/")

In [12]:
# train one epoch
from tqdm import tqdm

losses = []
for i, (x, y) in tqdm(enumerate(valid_dataloader)):
    optimizer.zero_grad()
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())

        for tag, value in model.named_parameters():
            if value.grad is not None:
                logger.add_histogram(tag + "/grad", value.grad.cpu(), i*10)

print('done 1 epoch', 'loss:', np.array(losses).mean())

0it [00:00, ?it/s]

iter 0 loss 18.12038803100586


10it [00:37,  3.51s/it]

iter 10 loss 19.25037332014604


20it [01:07,  3.55s/it]

iter 20 loss 20.473399389357795


30it [01:43,  3.68s/it]

iter 30 loss 21.38740570314469


40it [02:17,  3.30s/it]

iter 40 loss 22.757802498049852


50it [02:46,  3.34s/it]

done 1 epoch loss: 23.77406593322754


In [19]:
model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
model = initialization_using_word2vec(model)
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.000005)

wandb.init(
    # set the wandb project where this run will be logged
    project="bd4h-project",
    
    # track hyperparameters and run metadata
    config={
    "learning_rate": 0.000005,
    "architecture": "HLAN+LE",
    "dataset": "mimic-50",
    "epochs": 1,
    }
)

wandb.watch(model, criterion, log='all', log_freq=1)

logger = SummaryWriter("./logs/")

losses = []
for i, (x, y) in tqdm(enumerate(valid_dataloader)):
    optimizer.zero_grad()
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    wandb.log({'loss': loss.item()})

    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())

    for layer, (tag, value) in enumerate(model.named_parameters()):
        logger.add_histogram(f"grad/{layer}.{tag}", value.grad.cpu(), i)

print('done 1 epoch', 'loss:', np.array(losses).mean())


cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


0it [00:00, ?it/s]

iter 0 loss 19.893186569213867


10it [00:44,  4.20s/it]

iter 10 loss 22.45019635287198


20it [01:24,  3.94s/it]

iter 20 loss 24.543202173142205


30it [01:56,  3.22s/it]

iter 30 loss 25.426361330093876


40it [02:30,  3.26s/it]

iter 40 loss 26.752272256990757


50it [03:07,  3.74s/it]

done 1 epoch loss: 27.778998069763183


In [23]:
model.embeddings.weight.data.mean()

tensor(0.0014)

In [26]:
model.embeddings.weight.data

tensor([[ 6.1396e-03,  1.3556e-03, -3.6092e-04,  ..., -5.0547e-03,
         -3.1368e-03,  2.8457e-03],
        [ 2.0926e+00, -4.5464e-01,  1.8291e+00,  ..., -3.2387e+00,
          2.0335e+00,  1.9512e+00],
        [ 1.2300e+00,  1.4408e-01,  4.1557e-01,  ..., -1.0325e+00,
         -3.5222e-01,  1.2765e+00],
        ...,
        [-1.0968e-02, -9.3574e-03,  4.2765e-03,  ..., -9.8647e-03,
         -8.0069e-03, -9.9252e-03],
        [-1.9467e-02, -5.4372e-03,  3.0623e-03,  ...,  9.0115e-02,
         -6.0075e-02, -1.6655e-02],
        [-1.8207e-02,  7.9990e-03,  9.5015e-03,  ...,  3.4038e-02,
         -1.3660e-02, -8.8746e-03]])

In [29]:
model2 = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
model2 = initialization_using_word2vec(model2)
model2.embeddings.weight.data

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


tensor([[ 7.3915e-04,  4.0171e-03,  5.6544e-03,  ..., -4.6091e-03,
         -6.1196e-03,  8.1953e-04],
        [ 2.0924e+00, -4.5478e-01,  1.8291e+00,  ..., -3.2387e+00,
          2.0336e+00,  1.9510e+00],
        [ 1.2299e+00,  1.4384e-01,  4.1572e-01,  ..., -1.0323e+00,
         -3.5227e-01,  1.2763e+00],
        ...,
        [-1.0968e-02, -9.3574e-03,  4.2765e-03,  ..., -9.8647e-03,
         -8.0069e-03, -9.9252e-03],
        [-1.9467e-02, -5.4372e-03,  3.0623e-03,  ...,  9.0115e-02,
         -6.0075e-02, -1.6655e-02],
        [-1.8207e-02,  7.9990e-03,  9.5015e-03,  ...,  3.4038e-02,
         -1.3660e-02, -8.8746e-03]])

In [30]:
model2.embeddings.weight.data.mean()

tensor(0.0014)

In [35]:
model2.embeddings.weight.data[10,:]

tensor([-3.0873e+00, -1.2834e+00, -3.4314e+00,  2.7273e+00, -1.1918e+00,
        -1.5990e-01, -2.5308e+00,  4.2257e-01, -3.1923e-01,  2.4978e+00,
        -1.2897e-01,  1.7628e+00, -2.2354e-02,  4.0559e+00,  2.1190e-02,
         8.5813e-01,  8.9449e-01,  2.1671e-01,  4.0218e-01, -2.7727e-01,
         1.5968e+00, -1.1943e+00, -1.3494e+00,  2.7491e+00,  1.9064e+00,
        -4.4494e-01,  1.2601e-01,  1.3564e+00, -1.1144e+00,  5.0334e-01,
         1.3037e-01,  9.4918e-01,  3.2484e+00, -5.2825e-01, -6.0826e-03,
         1.9672e+00,  6.7840e-01,  1.4207e+00,  3.2294e+00,  1.4503e+00,
         3.3367e+00, -3.9044e-02, -2.7064e+00,  7.6135e-01,  2.6548e+00,
         1.3578e+00, -2.3968e+00,  1.9119e-01,  1.5119e+00, -8.1424e-01,
        -5.1204e-01, -1.4608e+00, -1.4920e+00,  1.6127e+00, -2.6102e+00,
        -1.9838e+00, -1.5403e+00, -2.4108e+00,  2.4294e+00,  5.9733e-01,
         7.6137e-01,  2.5147e+00, -5.3443e-01, -5.3333e-01, -3.2919e+00,
        -2.8149e+00, -2.2459e+00, -2.8667e-01,  4.5

In [36]:
model.embeddings.weight.data[10,:]

tensor([-3.0871e+00, -1.2833e+00, -3.4314e+00,  2.7275e+00, -1.1919e+00,
        -1.5964e-01, -2.5308e+00,  4.2277e-01, -3.1932e-01,  2.4977e+00,
        -1.2909e-01,  1.7629e+00, -2.2438e-02,  4.0560e+00,  2.0990e-02,
         8.5792e-01,  8.9459e-01,  2.1686e-01,  4.0219e-01, -2.7713e-01,
         1.5968e+00, -1.1944e+00, -1.3493e+00,  2.7493e+00,  1.9063e+00,
        -4.4497e-01,  1.2614e-01,  1.3563e+00, -1.1143e+00,  5.0354e-01,
         1.3013e-01,  9.4936e-01,  3.2485e+00, -5.2824e-01, -6.2785e-03,
         1.9672e+00,  6.7821e-01,  1.4209e+00,  3.2293e+00,  1.4501e+00,
         3.3366e+00, -3.9110e-02, -2.7065e+00,  7.6147e-01,  2.6547e+00,
         1.3579e+00, -2.3970e+00,  1.9140e-01,  1.5119e+00, -8.1431e-01,
        -5.1223e-01, -1.4610e+00, -1.4919e+00,  1.6128e+00, -2.6103e+00,
        -1.9840e+00, -1.5402e+00, -2.4107e+00,  2.4292e+00,  5.9744e-01,
         7.6122e-01,  2.5147e+00, -5.3418e-01, -5.3354e-01, -3.2918e+00,
        -2.8148e+00, -2.2456e+00, -2.8683e-01,  4.5

value of embeddings didn't change much after training

In [38]:
x, y = next(iter(train_dataloader))

In [43]:
x.min()

tensor(0)

In [44]:
y

tensor([[0., 0., 0.,  ..., 0., 1., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [1., 0., 0.,  ..., 0., 1., 1.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [1., 0., 0.,  ..., 0., 0., 0.]])

In [45]:
y.shape

torch.Size([32, 50])

In [46]:
y.sum(dim=1)

tensor([ 3.,  3.,  1.,  4.,  6.,  7.,  4.,  6., 10.,  4.,  9.,  5.,  6.,  2.,
         4.,  2.,  3.,  3.,  6.,  1.,  2., 13.,  5.,  3., 12.,  2.,  3.,  5.,
         1.,  9.,  3.,  2.])

we used the wrong loss function! Each y is a multi-hot label, so in stead of CrossEntropyLoss we need to use BCEWithLogitsLoss

In [47]:
model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
model = initialization_using_word2vec(model)
criterion = nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.000005)

wandb.init(
    # set the wandb project where this run will be logged
    project="bd4h-project",
    
    # track hyperparameters and run metadata
    config={
    "learning_rate": 0.000005,
    "architecture": "HLAN+LE",
    "dataset": "mimic-50",
    "epochs": 1,
    "loss": "BCEWithLogitsLoss"
    }
)

wandb.watch(model, criterion, log='all', log_freq=1)

logger = SummaryWriter("./logs/")

losses = []
for i, (x, y) in tqdm(enumerate(valid_dataloader)):
    optimizer.zero_grad()
    y_hat = model(x)
    loss = criterion(y_hat, y)
    loss.backward()
    losses.append(loss.item())
    optimizer.step()
    wandb.log({'loss': loss.item()})

    if i % 10 == 0:
        print('iter', i, 'loss', np.array(losses).mean())

    for layer, (tag, value) in enumerate(model.named_parameters()):
        logger.add_histogram(f"grad/{layer}.{tag}", value.grad.cpu(), i)

print('done 1 epoch', 'loss:', np.array(losses).mean())


cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


0it [00:00, ?it/s]

iter 0 loss 0.831281840801239


10it [00:39,  3.43s/it]

iter 10 loss 0.8375096754594282


20it [01:25,  4.29s/it]

iter 20 loss 0.8310460930778867


30it [01:59,  3.36s/it]

iter 30 loss 0.8223150814733198


40it [02:36,  3.83s/it]

iter 40 loss 0.8104217706657038


50it [03:10,  3.81s/it]

done 1 epoch loss: 0.7969258832931518


yay! now loss did go down during training!

# Overfit on a few data points

In [73]:
import torch
from torch.utils.data import DataLoader, Subset

def create_subset_dataloader(original_loader, num_samples=10):
    """
    Create a new DataLoader with a subset of the original data.

    Parameters:
    - original_loader (DataLoader): The original DataLoader.
    - num_samples (int): Number of samples to include in the subset.

    Returns:
    - DataLoader: A new DataLoader with the subset of data.
    """
    # Get the original dataset from the DataLoader
    original_dataset = original_loader.dataset

    # Check if the desired number of samples is not more than the dataset size
    num_samples = min(num_samples, len(original_dataset))

    # Create a subset
    subset_indices = torch.arange(num_samples)
    subset = Subset(original_dataset, subset_indices)

    # Create a new DataLoader with the subset
    subset_loader = DataLoader(subset, batch_size=original_loader.batch_size, shuffle=False)

    return subset_loader

# Example usage
# Assuming original_loader is your existing DataLoader
new_loader = create_subset_dataloader(train_dataloader, num_samples=1)

In [11]:
import sys
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.tensorboard import SummaryWriter
import numpy as np
from tqdm import tqdm
import wandb

# import constants
from constants import *

# Add the parent directory to the system path
sys.path.append("..")

from utils import create_vocabulary_label_pre_split, create_vocabulary
from utils import load_data_multilabel_pre_split , create_dataloaders, initialization_using_word2vec
from HAN_model import HierarchicalAttentionNetwork

from metrics import *


def train(dataloader, vocab_size, epochs=1, lr=0.0005, log=True):
    model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
    model = initialization_using_word2vec(model)
    criterion = nn.BCEWithLogitsLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min', patience=3)

    if log:
        wandb.init(
            # set the wandb project where this run will be logged
            project="bd4h-project",
            
            # track hyperparameters and run metadata
            config={
            "learning_rate": lr,
            "architecture": "HLAN+LE",
            "dataset": "mimic-50",
            "epochs": 1,
            "loss": "BCEWithLogitsLoss"
            }
        )
        wandb.watch(model, criterion, log='all', log_freq=1)

    for epoch in tqdm(range(epochs)):
        losses = []
        for x, y in dataloader:
            optimizer.zero_grad()
            y_hat = model(x)
            loss = criterion(y_hat, y)
            loss.backward()
            losses.append(loss.item())
            optimizer.step()
            
            if log:
                wandb.log({'loss': loss.item()})
                for layer, (tag, value) in enumerate(model.named_parameters()):
                    logger.add_histogram(f"grad/{layer}.{tag}", value.grad.cpu(), i)
            
        print('epoch', epoch, 'loss', np.array(losses).mean())
        scheduler.step(np.array(losses).mean())

    return model


In [50]:
from train import train

In [75]:
model = train(new_loader, vocab_size, epochs=100, lr=0.005, log=False)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


  3%|▎         | 3/100 [00:00<00:11,  8.09it/s]

epoch 0 loss 0.8193429708480835
epoch 1 loss 0.4199869632720947
epoch 2 loss 0.25279635190963745


  5%|▌         | 5/100 [00:00<00:09,  9.56it/s]

epoch 3 loss 0.17087604105472565
epoch 4 loss 0.13766346871852875
epoch 5 loss 0.11568380147218704


  9%|▉         | 9/100 [00:00<00:08, 10.65it/s]

epoch 6 loss 0.10351071506738663
epoch 7 loss 0.09372936934232712
epoch 8 loss 0.05058024078607559


 11%|█         | 11/100 [00:01<00:08, 10.95it/s]

epoch 9 loss 0.05220650136470795
epoch 10 loss 0.06845586001873016
epoch 11 loss 0.035004034638404846


 15%|█▌        | 15/100 [00:01<00:07, 11.27it/s]

epoch 12 loss 0.027773218229413033
epoch 13 loss 0.025817571207880974
epoch 14 loss 0.03080725483596325


 17%|█▋        | 17/100 [00:01<00:07, 10.85it/s]

epoch 15 loss 0.02734314650297165
epoch 16 loss 0.024270500987768173
epoch 17 loss 0.02255089208483696


 21%|██        | 21/100 [00:01<00:07, 11.10it/s]

epoch 18 loss 0.01995673216879368
epoch 19 loss 0.016502315178513527
epoch 20 loss 0.014533953741192818


 23%|██▎       | 23/100 [00:02<00:06, 11.20it/s]

epoch 21 loss 0.012399119324982166
epoch 22 loss 0.009690814651548862
epoch 23 loss 0.00843256525695324


 27%|██▋       | 27/100 [00:02<00:06, 10.72it/s]

epoch 24 loss 0.007241958752274513
epoch 25 loss 0.006641285493969917
epoch 26 loss 0.006502771284431219


 29%|██▉       | 29/100 [00:02<00:06, 10.89it/s]

epoch 27 loss 0.006486282218247652
epoch 28 loss 0.006434093229472637
epoch 29 loss 0.0063546388410031796


 33%|███▎      | 33/100 [00:03<00:06, 11.15it/s]

epoch 30 loss 0.006206745281815529
epoch 31 loss 0.006082253064960241
epoch 32 loss 0.006024539936333895


 35%|███▌      | 35/100 [00:03<00:05, 10.95it/s]

epoch 33 loss 0.006009091157466173
epoch 34 loss 0.0059487842954695225
epoch 35 loss 0.005876206327229738


 39%|███▉      | 39/100 [00:03<00:05, 10.61it/s]

epoch 36 loss 0.005800621118396521
epoch 37 loss 0.005745928268879652
epoch 38 loss 0.005706793628633022


 41%|████      | 41/100 [00:03<00:05, 10.65it/s]

epoch 39 loss 0.0056708138436079025
epoch 40 loss 0.0056336005218327045
epoch 41 loss 0.005591763649135828


 45%|████▌     | 45/100 [00:04<00:05, 10.93it/s]

epoch 42 loss 0.005544483195990324
epoch 43 loss 0.005497067701071501
epoch 44 loss 0.005453088786453009


 47%|████▋     | 47/100 [00:04<00:04, 11.05it/s]

epoch 45 loss 0.00541327428072691
epoch 46 loss 0.005376121960580349
epoch 47 loss 0.005340891424566507


 51%|█████     | 51/100 [00:04<00:04, 11.03it/s]

epoch 48 loss 0.005304434336721897
epoch 49 loss 0.005266610998660326
epoch 50 loss 0.0052277701906859875


 53%|█████▎    | 53/100 [00:04<00:04, 10.90it/s]

epoch 51 loss 0.0051885140128433704
epoch 52 loss 0.005149777512997389


 55%|█████▌    | 55/100 [00:05<00:04, 10.23it/s]

epoch 53 loss 0.005111894570291042
epoch 54 loss 0.00507454015314579
epoch 55 loss 0.005037682130932808


 59%|█████▉    | 59/100 [00:05<00:03, 10.56it/s]

epoch 56 loss 0.005001228302717209
epoch 57 loss 0.004965164698660374
epoch 58 loss 0.004929390270262957


 61%|██████    | 61/100 [00:05<00:03, 10.72it/s]

epoch 59 loss 0.004893905017524958
epoch 60 loss 0.004858673084527254
epoch 61 loss 0.004823667928576469


 65%|██████▌   | 65/100 [00:06<00:03, 10.96it/s]

epoch 62 loss 0.004788875114172697
epoch 63 loss 0.004754288122057915
epoch 64 loss 0.004719865974038839


 67%|██████▋   | 67/100 [00:06<00:02, 11.05it/s]

epoch 65 loss 0.004685614258050919
epoch 66 loss 0.004651515278965235
epoch 67 loss 0.004617564845830202


 71%|███████   | 71/100 [00:06<00:02, 10.86it/s]

epoch 68 loss 0.0045837704092264175
epoch 69 loss 0.004550142679363489
epoch 70 loss 0.004516692366451025


 73%|███████▎  | 73/100 [00:06<00:02, 10.87it/s]

epoch 71 loss 0.004483422264456749
epoch 72 loss 0.004450351465493441
epoch 73 loss 0.00441746786236763


 77%|███████▋  | 77/100 [00:07<00:02, 11.09it/s]

epoch 74 loss 0.0043847751803696156
epoch 75 loss 0.0043522659689188
epoch 76 loss 0.0043199616484344006


 79%|███████▉  | 79/100 [00:07<00:01, 11.16it/s]

epoch 77 loss 0.004287846852093935
epoch 78 loss 0.004255935084074736
epoch 79 loss 0.0042241946794092655


 81%|████████  | 81/100 [00:07<00:01, 11.08it/s]

epoch 80 loss 0.004192637279629707
epoch 81 loss 0.004161263816058636


 85%|████████▌ | 85/100 [00:07<00:01, 10.48it/s]

epoch 82 loss 0.00413003284484148
epoch 83 loss 0.004098784644156694
epoch 84 loss 0.004064661450684071


 87%|████████▋ | 87/100 [00:08<00:01, 10.18it/s]

epoch 85 loss 0.0040285480208694935
epoch 86 loss 0.003990148659795523


 89%|████████▉ | 89/100 [00:08<00:01,  9.93it/s]

epoch 87 loss 0.003945787902921438
epoch 88 loss 0.0038994222413748503


 92%|█████████▏| 92/100 [00:08<00:00, 10.10it/s]

epoch 89 loss 0.0038539450615644455
epoch 90 loss 0.0038117943331599236
epoch 91 loss 0.0037730804178863764


 94%|█████████▍| 94/100 [00:08<00:00, 10.41it/s]

epoch 92 loss 0.0037333103828132153
epoch 93 loss 0.003689634846523404
epoch 94 loss 0.003645260352641344


 98%|█████████▊| 98/100 [00:09<00:00, 10.39it/s]

epoch 95 loss 0.003603600896894932
epoch 96 loss 0.003564758924767375
epoch 97 loss 0.003527594730257988


100%|██████████| 100/100 [00:09<00:00, 10.64it/s]

epoch 98 loss 0.0034902903717011213
epoch 99 loss 0.0034513953141868114


In [82]:
x, y = next(iter(new_loader))
y_hat = model(x)
print(y.argmax(dim=1))
print(y_hat.argmax(dim=1))

tensor([6])
tensor([48])


In [86]:
y_hat

tensor([[-31.5819, -12.2134, -11.6357, -12.3696, -15.3418, -15.8302,   9.2258,
         -22.1400, -25.6424, -18.1387, -11.7790, -13.0992, -11.1329, -13.5590,
         -22.6651, -12.9194,  -8.6934, -12.4655, -18.7481, -21.6440, -11.1930,
         -11.3491, -11.6307,  -9.9779, -10.5721, -13.8085, -11.0767, -12.3561,
         -11.0915, -14.5768, -13.5439, -25.0359, -14.6869, -21.9130,  -9.2226,
         -10.2231,  -9.9000,   1.6860, -26.5688, -15.5023, -16.1322, -23.2063,
         -13.3837, -42.9854, -15.0577, -15.3846, -13.0794, -14.0496,  16.9560,
         -19.7838]], grad_fn=<AddBackward0>)

In [87]:
y

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]])

In [91]:
torch.where(torch.sigmoid(y_hat) > 0.01, torch.ones(y_hat.shape), torch.zeros(y_hat.shape))

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]])

we have successfully overfit one data point!

# overfit 10 data points

In [6]:
ten_loader = create_subset_dataloader(train_dataloader, num_samples=10)

In [13]:
valid_loader = create_subset_dataloader(valid_dataloader, num_samples=10)

In [20]:
model = train(ten_loader, valid_loader, vocab_size, epochs=5, lr=0.005, log=False, verbose=True)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True
Created directory to save model checkpoint at ../checkpoints/WfsQP


1it [00:00,  1.04it/s]00:00<?, ?it/s]
1it [00:00,  3.21it/s]
 20%|██        | 1/5 [00:01<00:05,  1.28s/it]

Epoch 0 train loss: 0.8429926633834839
Epoch 0 valid loss: 0.9579273462295532
Epoch 0 valid accuracy: 0.025974025974025976
Epoch 0 valid precision: 0.11764705882352941
Epoch 0 valid recall: 0.03225806451612903
Epoch 0 valid f1: 0.05063291139240506
--------------------------------------------------


1it [00:00,  1.16it/s]
1it [00:00,  3.26it/s]
 40%|████      | 2/5 [00:02<00:03,  1.22s/it]

Epoch 1 train loss: 0.8514472842216492
Epoch 1 valid loss: 0.9401633739471436
Epoch 1 valid accuracy: 0.02631578947368421
Epoch 1 valid precision: 0.125
Epoch 1 valid recall: 0.03225806451612903
Epoch 1 valid f1: 0.05128205128205128
--------------------------------------------------


1it [00:00,  1.19it/s]
1it [00:00,  3.31it/s]
 60%|██████    | 3/5 [00:03<00:02,  1.18s/it]

Epoch 2 train loss: 0.7192738056182861
Epoch 2 valid loss: 0.6289535760879517
Epoch 2 valid accuracy: 0.043478260869565216
Epoch 2 valid precision: 0.11764705882352941
Epoch 2 valid recall: 0.06451612903225806
Epoch 2 valid f1: 0.08333333333333333
--------------------------------------------------


1it [00:00,  1.17it/s]
1it [00:00,  3.34it/s]
 80%|████████  | 4/5 [00:04<00:01,  1.18s/it]

Epoch 3 train loss: 0.435390442609787
Epoch 3 valid loss: 0.7766780853271484
Epoch 3 valid accuracy: 0.09482758620689655
Epoch 3 valid precision: 0.16923076923076924
Epoch 3 valid recall: 0.1774193548387097
Epoch 3 valid f1: 0.1732283464566929
--------------------------------------------------


1it [00:00,  1.17it/s]
1it [00:00,  3.21it/s]
100%|██████████| 5/5 [00:05<00:00,  1.19s/it]

Epoch 4 train loss: 0.40000200271606445
Epoch 4 valid loss: 0.7848633527755737
Epoch 4 valid accuracy: 0.0859375
Epoch 4 valid precision: 0.14285714285714285
Epoch 4 valid recall: 0.1774193548387097
Epoch 4 valid f1: 0.15827338129496404
--------------------------------------------------


In [110]:
x, y = next(iter(ten_loader))
y_hat = model(x)
print(y.argmax(dim=1))

tensor([ 6,  0, 14,  8,  0,  0,  0,  2,  6,  8])


In [112]:
y

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
        [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [1., 0., 0., 1., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
       

In [113]:
torch.where(torch.sigmoid(y_hat) > 0.2, torch.ones(y_hat.shape), torch.zeros(y_hat.shape))

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
        [1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1.,
         0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
         0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [1., 0., 0., 1., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0.,
       

# Overfit 2 batches of data

In [115]:
two_batches_loader = create_subset_dataloader(train_dataloader, num_samples=64)

In [120]:
model = train(two_batches_loader, vocab_size, epochs=100, lr=0.005, log=True)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


  1%|          | 1/100 [00:11<18:33, 11.24s/it]

epoch 0 loss 0.6893196403980255


  2%|▏         | 2/100 [00:21<16:57, 10.39s/it]

epoch 1 loss 0.5394605845212936


  3%|▎         | 3/100 [00:30<15:56,  9.86s/it]

epoch 2 loss 0.4229678809642792


  4%|▍         | 4/100 [00:37<14:05,  8.81s/it]

epoch 3 loss 0.3586670905351639


  5%|▌         | 5/100 [00:44<12:57,  8.18s/it]

epoch 4 loss 0.31571295857429504


  6%|▌         | 6/100 [00:51<12:11,  7.78s/it]

epoch 5 loss 0.2897985577583313


  7%|▋         | 7/100 [01:00<12:39,  8.17s/it]

epoch 6 loss 0.26861102879047394


  8%|▊         | 8/100 [01:10<13:25,  8.75s/it]

epoch 7 loss 0.2534335255622864


  9%|▉         | 9/100 [01:18<12:48,  8.45s/it]

epoch 8 loss 0.24157759547233582


 10%|█         | 10/100 [01:26<12:32,  8.36s/it]

epoch 9 loss 0.22888939082622528


 11%|█         | 11/100 [01:33<11:42,  7.90s/it]

epoch 10 loss 0.2215374931693077


 12%|█▏        | 12/100 [01:39<10:59,  7.49s/it]

epoch 11 loss 0.2130671963095665


 13%|█▎        | 13/100 [01:49<11:50,  8.16s/it]

epoch 12 loss 0.20530206710100174


 14%|█▍        | 14/100 [01:59<12:30,  8.73s/it]

epoch 13 loss 0.206512913107872


 15%|█▌        | 15/100 [02:07<12:07,  8.56s/it]

epoch 14 loss 0.20684615522623062


 16%|█▌        | 16/100 [02:15<11:27,  8.19s/it]

epoch 15 loss 0.20085904002189636


 17%|█▋        | 17/100 [02:22<11:06,  8.03s/it]

epoch 16 loss 0.19995411485433578


 18%|█▊        | 18/100 [02:30<10:53,  7.96s/it]

epoch 17 loss 0.20205532014369965


 19%|█▉        | 19/100 [02:40<11:33,  8.56s/it]

epoch 18 loss 0.19636888056993484


 20%|██        | 20/100 [02:47<10:49,  8.11s/it]

epoch 19 loss 0.1864580661058426


 21%|██        | 21/100 [02:57<11:21,  8.63s/it]

epoch 20 loss 0.1801273375749588


 22%|██▏       | 22/100 [03:08<12:11,  9.38s/it]

epoch 21 loss 0.18123611062765121


 23%|██▎       | 23/100 [03:19<12:28,  9.72s/it]

epoch 22 loss 0.1774047240614891


 24%|██▍       | 24/100 [03:26<11:32,  9.12s/it]

epoch 23 loss 0.16386164724826813


 25%|██▌       | 25/100 [03:35<11:19,  9.06s/it]

epoch 24 loss 0.16001787036657333


 26%|██▌       | 26/100 [03:43<10:36,  8.60s/it]

epoch 25 loss 0.1528056636452675


 27%|██▋       | 27/100 [03:49<09:45,  8.02s/it]

epoch 26 loss 0.14428751170635223


 28%|██▊       | 28/100 [03:56<09:07,  7.61s/it]

epoch 27 loss 0.13980965316295624


 29%|██▉       | 29/100 [04:05<09:28,  8.01s/it]

epoch 28 loss 0.13516875356435776


 30%|███       | 30/100 [04:13<09:14,  7.92s/it]

epoch 29 loss 0.13543881103396416


 31%|███       | 31/100 [04:20<08:58,  7.80s/it]

epoch 30 loss 0.1335744671523571


 32%|███▏      | 32/100 [04:28<08:44,  7.72s/it]

epoch 31 loss 0.1314801201224327


 33%|███▎      | 33/100 [04:37<09:00,  8.07s/it]

epoch 32 loss 0.13116095960140228


 34%|███▍      | 34/100 [04:43<08:26,  7.67s/it]

epoch 33 loss 0.14098674803972244


 35%|███▌      | 35/100 [04:50<08:00,  7.39s/it]

epoch 34 loss 0.15085270255804062


 36%|███▌      | 36/100 [04:57<07:36,  7.13s/it]

epoch 35 loss 0.15590766072273254


 37%|███▋      | 37/100 [05:04<07:41,  7.32s/it]

epoch 36 loss 0.1643742471933365


 38%|███▊      | 38/100 [05:11<07:22,  7.14s/it]

epoch 37 loss 0.1507372260093689


 39%|███▉      | 39/100 [05:21<08:04,  7.95s/it]

epoch 38 loss 0.14343036711215973


 40%|████      | 40/100 [05:28<07:33,  7.56s/it]

epoch 39 loss 0.1384417936205864


 41%|████      | 41/100 [05:36<07:48,  7.94s/it]

epoch 40 loss 0.13463637605309486


 42%|████▏     | 42/100 [05:44<07:29,  7.75s/it]

epoch 41 loss 0.13242574408650398


 43%|████▎     | 43/100 [05:52<07:27,  7.84s/it]

epoch 42 loss 0.13204479590058327


 44%|████▍     | 44/100 [05:59<07:00,  7.51s/it]

epoch 43 loss 0.13157068192958832


 45%|████▌     | 45/100 [06:09<07:33,  8.25s/it]

epoch 44 loss 0.13104566931724548


 46%|████▌     | 46/100 [06:18<07:47,  8.66s/it]

epoch 45 loss 0.13041288033127785


 47%|████▋     | 47/100 [06:25<07:08,  8.09s/it]

epoch 46 loss 0.12993425503373146


 48%|████▊     | 48/100 [06:32<06:45,  7.79s/it]

epoch 47 loss 0.1293630227446556


 49%|████▉     | 49/100 [06:39<06:18,  7.42s/it]

epoch 48 loss 0.12877492606639862


 50%|█████     | 50/100 [06:45<05:57,  7.14s/it]

epoch 49 loss 0.1282314620912075


 51%|█████     | 51/100 [06:51<05:38,  6.91s/it]

epoch 50 loss 0.1277257800102234


 52%|█████▏    | 52/100 [06:58<05:26,  6.79s/it]

epoch 51 loss 0.12698376178741455


 53%|█████▎    | 53/100 [07:04<05:14,  6.70s/it]

epoch 52 loss 0.1263667233288288


 54%|█████▍    | 54/100 [07:11<05:04,  6.63s/it]

epoch 53 loss 0.12590481713414192


 55%|█████▌    | 55/100 [07:17<04:54,  6.55s/it]

epoch 54 loss 0.12550506368279457


 56%|█████▌    | 56/100 [07:26<05:13,  7.13s/it]

epoch 55 loss 0.12513266876339912


 57%|█████▋    | 57/100 [07:33<05:13,  7.29s/it]

epoch 56 loss 0.1248079165816307


 58%|█████▊    | 58/100 [07:40<04:54,  7.02s/it]

epoch 57 loss 0.12444084882736206


 59%|█████▉    | 59/100 [07:46<04:39,  6.82s/it]

epoch 58 loss 0.12409771233797073


 60%|██████    | 60/100 [07:53<04:32,  6.82s/it]

epoch 59 loss 0.12373586371541023


 61%|██████    | 61/100 [08:02<04:51,  7.47s/it]

epoch 60 loss 0.12335054203867912


 62%|██████▏   | 62/100 [08:09<04:42,  7.42s/it]

epoch 61 loss 0.1229386068880558


 63%|██████▎   | 63/100 [08:17<04:41,  7.61s/it]

epoch 62 loss 0.12250563502311707


 64%|██████▍   | 64/100 [08:24<04:22,  7.28s/it]

epoch 63 loss 0.12210097536444664


 65%|██████▌   | 65/100 [08:32<04:27,  7.65s/it]

epoch 64 loss 0.12174494192004204


 66%|██████▌   | 66/100 [08:42<04:36,  8.14s/it]

epoch 65 loss 0.12149905413389206


 67%|██████▋   | 67/100 [08:51<04:39,  8.48s/it]

epoch 66 loss 0.12127004936337471


 68%|██████▊   | 68/100 [08:58<04:21,  8.16s/it]

epoch 67 loss 0.12104660272598267


 69%|██████▉   | 69/100 [09:06<04:05,  7.92s/it]

epoch 68 loss 0.12081402912735939


 70%|███████   | 70/100 [09:12<03:43,  7.46s/it]

epoch 69 loss 0.12057529762387276


 71%|███████   | 71/100 [09:19<03:27,  7.17s/it]

epoch 70 loss 0.1203366369009018


 72%|███████▏  | 72/100 [09:27<03:32,  7.61s/it]

epoch 71 loss 0.12009390816092491


 73%|███████▎  | 73/100 [09:34<03:15,  7.25s/it]

epoch 72 loss 0.1198529377579689


 74%|███████▍  | 74/100 [09:40<03:01,  7.00s/it]

epoch 73 loss 0.11963000148534775


 75%|███████▌  | 75/100 [09:49<03:10,  7.62s/it]

epoch 74 loss 0.11942493915557861


 76%|███████▌  | 76/100 [09:58<03:14,  8.10s/it]

epoch 75 loss 0.11921478062868118


 77%|███████▋  | 77/100 [10:06<03:04,  8.01s/it]

epoch 76 loss 0.11899644881486893


 78%|███████▊  | 78/100 [10:12<02:45,  7.52s/it]

epoch 77 loss 0.11875937506556511


 79%|███████▉  | 79/100 [10:19<02:30,  7.19s/it]

epoch 78 loss 0.11850178614258766


 80%|████████  | 80/100 [10:28<02:35,  7.77s/it]

epoch 79 loss 0.11824177205562592


 81%|████████  | 81/100 [10:37<02:34,  8.15s/it]

epoch 80 loss 0.11804034188389778


 82%|████████▏ | 82/100 [10:43<02:17,  7.61s/it]

epoch 81 loss 0.11784709990024567


 83%|████████▎ | 83/100 [10:50<02:03,  7.28s/it]

epoch 82 loss 0.117656409740448


 84%|████████▍ | 84/100 [10:59<02:06,  7.92s/it]

epoch 83 loss 0.11746703833341599


 85%|████████▌ | 85/100 [11:07<01:56,  7.76s/it]

epoch 84 loss 0.117281474173069


 86%|████████▌ | 86/100 [11:16<01:56,  8.31s/it]

epoch 85 loss 0.11709403246641159


 87%|████████▋ | 87/100 [11:26<01:53,  8.75s/it]

epoch 86 loss 0.11690415814518929


 88%|████████▊ | 88/100 [11:36<01:47,  8.98s/it]

epoch 87 loss 0.11671089753508568


 89%|████████▉ | 89/100 [11:45<01:40,  9.13s/it]

epoch 88 loss 0.11651740223169327


 90%|█████████ | 90/100 [11:54<01:29,  9.00s/it]

epoch 89 loss 0.11632440984249115


 91%|█████████ | 91/100 [12:01<01:16,  8.48s/it]

epoch 90 loss 0.11613336205482483


 92%|█████████▏| 92/100 [12:08<01:05,  8.16s/it]

epoch 91 loss 0.11594498157501221


 93%|█████████▎| 93/100 [12:15<00:53,  7.64s/it]

epoch 92 loss 0.11575526744127274


 94%|█████████▍| 94/100 [12:22<00:45,  7.59s/it]

epoch 93 loss 0.115562304854393


 95%|█████████▌| 95/100 [12:30<00:37,  7.58s/it]

epoch 94 loss 0.11536028236150742


 96%|█████████▌| 96/100 [12:37<00:30,  7.51s/it]

epoch 95 loss 0.11515707895159721


 97%|█████████▋| 97/100 [12:45<00:22,  7.50s/it]

epoch 96 loss 0.11494201421737671


 98%|█████████▊| 98/100 [12:54<00:16,  8.09s/it]

epoch 97 loss 0.11468945443630219


 99%|█████████▉| 99/100 [13:01<00:07,  7.62s/it]

epoch 98 loss 0.1144183985888958


100%|██████████| 100/100 [13:07<00:00,  7.88s/it]

epoch 99 loss 0.11418961361050606


# Overfit to step 600

when training with full data, loss seems to blow up at the end of an epoch, near step 250. Let's triple the data size and see if the same happen to smaller dataset

In [8]:
from utils import create_subset_dataloader
dataloader = create_subset_dataloader(train_dataloader, num_samples=64)

In [11]:
model = train(dataloader, vocab_size, epochs=300, lr=0.005, log=True, verbose=False)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


2it [00:12,  6.21s/it] [00:00<?, ?it/s]
2it [00:10,  5.44s/it] [00:12<1:01:54, 12.42s/it]
2it [00:09,  4.79s/it] [00:23<57:11, 11.52s/it]  
2it [00:09,  4.97s/it] [00:32<52:39, 10.64s/it]
2it [00:09,  4.75s/it] [00:42<51:06, 10.36s/it]
2it [00:09,  4.81s/it] [00:52<49:25, 10.05s/it]
2it [00:10,  5.34s/it] [01:01<48:33,  9.91s/it]
2it [00:09,  4.77s/it] [01:12<49:38, 10.16s/it]
2it [00:09,  4.58s/it] [01:22<48:29,  9.97s/it]
2it [00:09,  4.67s/it] [01:31<47:06,  9.71s/it]
2it [00:11,  5.52s/it]0 [01:40<46:24,  9.60s/it]
2it [00:09,  4.67s/it]0 [01:51<48:22, 10.04s/it]
2it [00:09,  4.86s/it]0 [02:01<47:11,  9.83s/it]
2it [00:09,  4.81s/it]0 [02:10<46:52,  9.80s/it]
2it [00:10,  5.31s/it]0 [02:20<46:27,  9.75s/it]
2it [00:09,  4.67s/it]0 [02:31<47:33, 10.01s/it]
2it [00:08,  4.03s/it]0 [02:40<46:26,  9.81s/it]
2it [00:06,  3.40s/it]0 [02:48<43:48,  9.29s/it]
2it [00:08,  4.10s/it]0 [02:55<40:07,  8.54s/it]
2it [00:08,  4.18s/it]0 [03:03<39:31,  8.44s/it]
2it [00:06,  3.42s/it]0 [03:11<39:

all look good. weird that full data training has loss going up. Trying again with shuffled data and high lr

# Train for real

In [106]:
import sys
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.tensorboard import SummaryWriter
import numpy as np
from tqdm import tqdm
import wandb

# import constants
from constants import *

# Add the parent directory to the system path
sys.path.append("..")

from HLAN.data_util_gensim import create_vocabulary_label_pre_split, create_vocabulary
from utils import load_data_multilabel_pre_split , create_dataloaders, initialization_using_word2vec
from HAN_model import HierarchicalAttentionNetwork

from metrics import *


def train(dataloader, vocab_size, epochs=1, lr=0.0005, log=True, project="bd4h-project"):
    model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
    model = initialization_using_word2vec(model)
    criterion = nn.BCEWithLogitsLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min', patience=3)

    if log:
        wandb.init(
            # set the wandb project where this run will be logged
            project=project,
            
            # track hyperparameters and run metadata
            config={
            "learning_rate": lr,
            "architecture": "HLAN+LE",
            "dataset": "mimic-50",
            "epochs": epochs,
            "loss": "BCEWithLogitsLoss"
            }
        )
        wandb.watch(model, criterion, log='all', log_freq=1)

    logger = SummaryWriter("./logs/")

    for epoch in tqdm(range(epochs)):
        losses = []
        for x, y in dataloader:
            optimizer.zero_grad()
            y_hat = model(x)
            loss = criterion(y_hat, y)
            loss.backward()
            losses.append(loss.item())
            optimizer.step()
            
            if log:
                wandb.log({'loss': loss.item()})
                for layer, (tag, value) in enumerate(model.named_parameters()):
                    logger.add_histogram(f"grad/{layer}.{tag}", value.grad.cpu(), i)
            
        print('epoch', epoch, 'loss', np.array(losses).mean())
        scheduler.step(np.array(losses).mean())

    return model


In [93]:
model = train(train_dataloader, vocab_size, epochs=5, lr=0.00005, log=True)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


 20%|██        | 1/5 [13:39<54:37, 819.47s/it]

epoch 0 loss 0.38444908982209064


 40%|████      | 2/5 [26:26<39:25, 788.34s/it]

epoch 1 loss 0.39879510902133386


 60%|██████    | 3/5 [40:23<27:01, 810.68s/it]

epoch 2 loss 0.3494505654917404


 80%|████████  | 4/5 [53:56<13:31, 811.80s/it]

epoch 3 loss 0.34063784954811743


100%|██████████| 5/5 [1:07:04<00:00, 804.86s/it]

epoch 4 loss 0.3346067282404353


In [94]:
x, y = next(iter(new_loader))

In [95]:
y

tensor([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]])

In [100]:
y_hat = model(x)
torch.where(torch.sigmoid(y_hat) > 0.1, torch.ones(y_hat.shape), torch.zeros(y_hat.shape))

tensor([[1., 1., 1., 1., 0., 1., 0., 0., 1., 1., 1., 1., 0., 1., 1., 1., 0., 1.,
         1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 0., 0., 1., 1., 0., 0., 1., 0.,
         0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 1., 0., 0.]])

In [101]:
x

tensor([[41, 43, 22,  ...,  0,  0,  0]])

In [104]:
x[0,-30:]

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0])

#### train with higher initial learning rate

In [114]:
model = train(train_dataloader, vocab_size, epochs=5, lr=0.005, log=True)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


 20%|██        | 1/5 [13:35<54:23, 815.78s/it]

epoch 0 loss 0.3926349367077643


 40%|████      | 2/5 [27:03<40:33, 811.32s/it]

epoch 1 loss 0.38670202512514923


 60%|██████    | 3/5 [40:58<27:23, 821.70s/it]

epoch 2 loss 0.4673087099795285


 80%|████████  | 4/5 [55:09<13:53, 833.64s/it]

epoch 3 loss 0.3527893528283349


100%|██████████| 5/5 [1:09:44<00:00, 836.96s/it]

epoch 4 loss 0.34148601904923737


# Train full data 20 epochs

In [14]:
model = train(train_dataloader, vocab_size, epochs=20, lr=0.005, log=True, verbose=False)

cache_path: ../cache_vocabulary_label_pik/mimic3-ds-50-HAN_word_vocabulary.pik file_exists: True


wandb: WARNING Source type is set to 'repo' but some required information is missing from the environment. A job will not be created from this run. See https://docs.wandb.ai/guides/launch/create-job


253it [50:00, 11.86s/it]0:00<?, ?it/s]
253it [1:03:29, 15.06s/it]00<15:50:04, 3000.24s/it]
253it [1:10:56, 16.83s/it]3:29<17:22:50, 3476.12s/it]
253it [16:57,  4.02s/it]:04:26<18:05:54, 3832.62s/it]
253it [19:17,  4.57s/it]:21:24<12:05:42, 2721.41s/it]
253it [19:14,  4.56s/it]:40:41<8:59:20, 2157.37s/it] 
253it [21:49,  5.17s/it]:59:56<7:03:50, 1816.45s/it]
253it [1:21:40, 19.37s/it]1:45<5:57:38, 1650.64s/it]
253it [1:23:24, 19.78s/it]3:25<8:57:01, 2685.13s/it]
253it [56:57, 13.51s/it]:06:49<10:25:11, 3410.13s/it]
253it [35:01,  8.30s/it]8:03:47<9:28:43, 3412.40s/it]
253it [14:21,  3.41s/it]8:38:48<7:31:39, 3011.05s/it]
253it [15:58,  3.79s/it]8:53:10<5:14:17, 2357.23s/it]
253it [15:04,  3.57s/it]9:09:08<3:45:35, 1933.59s/it]
253it [15:30,  3.68s/it]9:24:13<2:42:15, 1622.66s/it]
253it [15:14,  3.61s/it]9:39:43<1:57:49, 1413.88s/it]
253it [18:51,  4.47s/it]9:54:57<1:24:13, 1263.47s/it]
253it [14:41,  3.48s/it]10:13:48<1:01:11, 1223.74s/it]
253it [15:34,  3.69s/it]10:28:30<37:21, 1120.93

In [15]:
torch.save(model.state_dict(), "./checkpoints/HLAN_LE_20epochs.pt")

let's evaluate the model. Here we need to use sigmoid activation because it is multiclass

In [ ]:
model = HierarchicalAttentionNetwork(vocab_size=vocab_size, embed_size=EMBED_SIZE, hidden_size=HIDDEN_SIZE, num_sentences=NUM_SENTENCES, sentence_length=SENTENCE_LENGTH, num_classes=NUM_CLASSES)
model.load_state_dict(torch.load("./checkpoints/HLAN_LE_20epochs.pt"))
model.eval()

In [52]:
all_logits = []
all_probs = []
all_labels = []
all_labels_actuals = []
with torch.no_grad():
    for x, y in tqdm(valid_dataloader):
        logits = model(x)
        probs = F.sigmoid(logits)
        labels = torch.where(probs > 0.5, torch.ones(probs.shape), torch.zeros(probs.shape))

        all_logits.append(list(logits.numpy()))
        all_probs.append(list(probs.numpy()))
        all_labels.append(list(labels.numpy()))
        all_labels_actuals.append(list(y.numpy()))

  0%|          | 0/50 [00:00<?, ?it/s]

100%|██████████| 50/50 [00:55<00:00,  1.11s/it]


In [55]:
thresholds = [0.1, 0.2, 0.3, 0.4, 0.5]
for threshold in thresholds:
    preds = np.concatenate(all_probs)
    actuals = np.concatenate(all_labels_actuals)
    preds = np.where(preds > threshold, 1, 0)
    acc, prec, rec, f1 = all_micro(preds.ravel(), actuals.ravel())
    print(f"Threshold: {threshold}")
    print(f"Accuracy: {acc}")
    print(f"Precision: {prec}")
    print(f"Recall: {rec}")
    print(f"F1: {f1}")
    print()

Threshold: 0.1
Accuracy: 0.18715774843641417
Precision: 0.20143605086013464
Recall: 0.725304319724227
F1: 0.3153039243233118

Threshold: 0.2
Accuracy: 0.22248635798800584
Precision: 0.30860311750599523
Recall: 0.44360659269632663
F1: 0.3639899235426703

Threshold: 0.3
Accuracy: 0.21055350553505536
Precision: 0.4007022471910112
Recall: 0.3073359905203059
F1: 0.34786319575687374

Threshold: 0.4
Accuracy: 0.16571075619871173
Precision: 0.47810590631364563
Recall: 0.20230528923839275
F1: 0.284308530769813

Threshold: 0.5
Accuracy: 0.11634076839933183
Precision: 0.5697786333012512
Recall: 0.12754497468490789
F1: 0.20843235630666315



the paper reached 64% F1 after 100 epochs. Ours reached 36% F1 after 30 epochs. Would 70 more epoch help?